ARrehab/ARrehab/Minigames/FaceGame/FaceViewController.swift

```diff
@@ -0,0 +1,358 @@
+//
+//  ARFaceViewController.swift
+//  ARrehab
+//
+//  Created by Alan Zhang on 2020/6/6.
+//  Copyright © 2020 Eric Wang. All rights reserved.
+//
+
+import Foundation
+import UIKit
+import ARKit
+
+
+/**
+ Face Minigame ViewController
+ */
+class FaceViewController: UIViewController, ARSessionDelegate {
+    
+    //changing elements
+    @IBOutlet weak var actionLabel: UILabel!
+    @IBOutlet weak var timeLeftBar: UIProgressView!
+    @IBOutlet weak var pointsLeftLabel: UILabel!
+    @IBOutlet weak var totalPointsLabel: UILabel!
+    @IBOutlet weak var livesLeftLabel: UILabel!
+    
+    //static elements
+    @IBOutlet weak var livesLeftTitleLabel: UILabel!
+    
+    //action buttons
+    @IBOutlet weak var startGameButton: UIButton!
+    
+    //views
+    @IBOutlet weak var sceneView: SCNView!
+    @IBOutlet weak var faceImageView: UIImageView!
+    
+    let session = ARSession()
+    var maskNode: Mask?
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        self.sceneView.backgroundColor = .clear
+        self.sceneView.scene = SCNScene()
+        self.sceneView.rendersContinuously = true
+        
+        // floating mask node
+        if let device = MTLCreateSystemDefaultDevice(), let geo = ARSCNFaceGeometry(device: device) {
+            self.maskNode = Mask(geometry: geo)
+            self.sceneView.scene?.rootNode.addChildNode(self.maskNode!)
+            self.maskNode?.position = SCNVector3(x: 0.0, y: 0.0, z: 0.0)
+        }
+        
+        // configure our ARKit tracking session for facial recognition
+        let config = ARFaceTrackingConfiguration()
+        config.worldAlignment = .gravity
+        session.delegate = self
+        session.run(config, options: [])
+        
+        self.updateUI()
+    }
+    
+    
+    // MARK: - User Actions
+    
+    @IBAction func startGame(_ sender: Any) {
+        self.startGame()
+    }
+    
+    
+    // MARK: - AR Session
+    
+    var currentFaceAnchor: ARFaceAnchor?
+    var currentFrame: ARFrame?
+    
+    func session(_ session: ARSession, didUpdate frame: ARFrame) {
+        self.currentFrame = frame
+        DispatchQueue.main.async {
+            // need to call heart beat on main thread
```

Comment:
what do you mean heart beat on main thread?

Reply:
Not my comment neither, will need to take a closer look